dmscripts/upload_dos_opportunities_email_list.py

```diff
@@ -0,0 +1,46 @@
+from dmscripts.helpers.supplier_data_helpers import SupplierFrameworkData
+
+
+def find_user_emails(supplier_users, services):
+    """Return email addresses for any supplier who has a service in services."""
+    email_addresses = []
+    for service in services:
+        for user in supplier_users.get(service['supplierId'], []):
+            email_addresses.append(user['email address'])
+    return email_addresses
```

Comment:
Essentially, with the imperative way of doing it, you're commanding a little robot with little instructions of what to do. With the functional way of doing it, you're more *describing the result* you want to achieve. As such it's also far easier for the next developer coming along to glean the programmer's intention.